### PR TITLE
locksmithctl/locksmithcl: fix endpoints resilience

### PR DESCRIPTION
**Pull request by tormath1**
with the recent etcd upgrade (https://github.com/kinvolk/locksmith/commit/1b5eb50),
the endpoints resilience were not ensured -> having failing endpoints
would make the `locksmithcl` commands fail.

in this commit, we patch the software (and not the vendor), to reproduce
a kind of resilience: we loop over the endpoints - if one of them is not
reachable we continue to the over.
The loop could have been done elsewhere in the `Create` command, but we
don't have access to the `client.httpKeysAPI` which holds the config
endpoints.

Signed-off-by: Mathieu Tortuyaux <mathieu@kinvolk.io>

See also: https://github.com/kinvolk/coreos-overlay/pull/1161

### CODE REVIEW

**Review by invidian — COMMENTED** [review]
@tormath1 maybe we could try upgrading to latest etcd client version, maybe there was a regression which is now fixed? I know there has been a lot of internal changes in etcd v3.5.0, which makes it much easier to consume via go modules.

**Comment by invidian** [comment]
Now it came to my mind, this should be quite easy to have integration tests, perhaps we can then ensure the desired behavior.

**Comment by tormath1** [comment]
> maybe we could try upgrading to latest etcd client version, maybe there was a regression which is now fixed? I know there has been a lot of internal changes in etcd v3.5.0, which makes it much easier to consume via go modules.

@invidian I spent quite some times to figure out the changes on the concerned files - but yeah maybe the issue has been fixed deeper elsewhere. I'll give a try to the upgrade - would be nice also in order to prepare any v3 migration.

File: locksmithctl/locksmithctl.go
Comment (invidian) [inline locksmithctl/locksmithctl.go]: It should be fine to use `fmt.Errorf` here, even though we don't wrap anything, to avoid importing `errors` :smile: 
```suggestion
	return nil, fmt.Errorf("unable to create an etcd client")
```

Comment (tormath1) [reply locksmithctl/locksmithctl.go]: Yeah for sure... I just don't know why there is no `fmt.Error` only :D

**Comment by tormath1** [comment]
For the record, I upgraded `etcd` to `/v2` - we still have the same behavior (see: https://github.com/kinvolk/locksmith/commits/tormath1/upgrade-etcd)

File: locksmithctl/locksmithctl.go
Comment (tormath1) [reply locksmithctl/locksmithctl.go]: I just checked, we actually import the `errors` to assert the errors is a `connection refused` one (https://github.com/kinvolk/locksmith/commit/b37d41de9e44e84a3d69e3b78769de75e15d87a7#diff-7b29e35460d650d4b39ff1302e2c54d5817c848ab7a497d19569e2329fbb8cdaR238)

**Comment by invidian** [comment]
>For the record, I upgraded etcd to /v2 - we still have the same behavior (see: https://github.com/kinvolk/locksmith/commits/tormath1/upgrade-etcd)

And what about v3 client?

**Comment by tormath1** [comment]
@invidian I started to have a look last Thursday - there are two things to considere:

- etcd/v3 client does not seem compatible with v2 server - so it's not a "simple" refactoring since we could [break](https://etcd.io/docs/v3.4/op-guide/v2-migration/) users with etcd/v2 nodes ? I agree we should upgrade to v3 but without a test cover and a way to provide a seamless transition it sounds a bit ... perilous >< 
- this interface: https://github.com/kinvolk/locksmith/blob/c96ae6e242605a6bdb8969faed1bd848d724e7e1/lock/etcd.go#L36-L42 does not seem to exist anymore with v3

**Comment by invidian** [comment]
Thanks for checking @tormath1. You're perhaps right about migration to v3, it doesn't seem trivial, if the data is stored in a different place. Meaning upgrading only some clients to v3 can cause multiple nodes to be updated simultaneously. Perhaps we could create an issue for this.

**Comment by tormath1** [comment]
@invidian I gave a deeper try to etcd/v3 upgrade - it compiles; let's run some tests to see the damages haha.

EDIT:
```
core@localhost ~ $ ./locksmithctl status
Available: 1
Max: 1
core@localhost ~ $ ./locksmithctl set-max 1234
Old-Max: 1
Max: 1234
core@localhost ~ $ ./locksmithctl status
Available: 1
Max: 1
```
works fine haha

**Comment by tormath1** [comment]
closed in favor of: https://github.com/kinvolk/locksmith/pull/12

**Comment by tormath1** [comment]
@kinvolk/flatcar-maintainers @invidian let's bring back this PR to life.

Since the on-going PR (https://github.com/kinvolk/locksmith/pull/12) to upgrade to etcd/v3 will take some time (need to properly test it, rewrite a few things); let's patch the current locksmith with the following changes to ship `locksmith` in the OS.

File: locksmithctl/locksmithctl.go
Comment (invidian) [inline locksmithctl/locksmithctl.go]:
```suggestion
			return nil, fmt.Errorf("creating etcd client: %w", err)
```

File: locksmithctl/locksmithctl.go
Comment (invidian) [inline locksmithctl/locksmithctl.go]:
```suggestion
			return nil, fmt.Errorf("creating etcd lock client: %w", err)
```

File: locksmithctl/locksmithctl.go
Comment (invidian) [inline locksmithctl/locksmithctl.go]:
```suggestion
	return nil, errors.New("no etcd endpoints available, tried: %s", strings.Join(",", globalFlags.Endpoints))
```

File: locksmithctl/locksmithctl.go
Comment (invidian) [inline locksmithctl/locksmithctl.go:226]: Can we have a comment somewhere about v2 client, kola tests and that this won't be needed with client v3?

**Review by invidian — CHANGES_REQUESTED** [review]
I suggested improvements to returned error messages and adding a comment to describe the situation, otherwise looks good :+1:

**Review by krnowak — COMMENTED** [review]
Looks good, but the commit message is confusing. What `Create` command? I don't see locksmithctl having a create command. You probably meant something about etcd internal create command or something?

**Comment by tormath1** [comment]
@krnowak right, the `Create` actually refers to the `Create` **method** in the `KeysAPI` interface but it's confusing and I don't remember what I wanted to say - I'll definitely reword the commit.

**Comment by tormath1** [comment]
@invidian @krnowak thanks for your review - comments have been adressed.

New body's commit message will be:
```
with the recent etcd upgrade (1b5eb50),
the endpoints resilience were not ensured -> having failing endpoints
would make the `locksmithcl` commands fail.

in this commit, we patch the software (and not the vendor), to reproduce
a kind of resilience: we loop over the endpoints - if one of them is not
reachable we continue to the over.
this patch can be reverted once the upgrade to etcd/v3 will be done.
```

**Review by krnowak — APPROVED** [review]

**Review by invidian — APPROVED** [review]
Looks fantastic! Great work @tormath1 :tada:

**Review by invidian — CHANGES_REQUESTED** [review]
One thing, can we squash commits before merging?

**Comment by tormath1** [comment]
@invidian obviously I'll squash the `fixup!` commits before merging if this is what you're talking about. :partying_face:

**Review by invidian — APPROVED** [review]

**Pull request merged**
